base-containers/kvm/Dockerfile

```diff
@@ -0,0 +1,102 @@
+ARG   	VERSION=latest
+ARG	REGISTRY
+
+# Rockylinux does not enable 9P virtfs in the shipped QEMU builds, hence we have to rebuild our own version
+# Inspired from https://github.com/acudovs/qemu-kvm-virtfs/blob/master/rpmbuild/build
+FROM rockylinux:8 as builder
+
+# Enable additional repos
+RUN 	dnf -y install 'dnf-command(config-manager)' &&\
+   	dnf config-manager --set-enabled powertools
+
+# Install dependencies to build QEMU
+RUN 	yum -y update && yum -y install \
+        glusterfs-api-devel \
+        glusterfs-devel \
+        iasl \
+        libcacard-devel \
+	libpmem-devel \
+	nss-devel \
+	pkgconfig \
+	spice-protocol \
+	spice-server-devel \
+	usbredir-devel \
+	yum-utils \
+	'pkgconfig(epoxy)' \
+	'pkgconfig(gbm)' \
+	'pkgconfig(libdrm)' \
+	git
+
+# Apply patches and build QEMU
+WORKDIR /opt
+RUN 	yum-builddep -y qemu-kvm
+#RUN 	yumdownloader --source qemu-kvm
+RUN	wget "http://download.rockylinux.org/pub/rocky/8/AppStream/source/tree/Packages/q/qemu-kvm-6.2.0-32.module%2Bel8.8.0%2B1279%2B230c2115.src.rpm"
+RUN 	rpm -Uvh qemu-kvm-*.src.rpm
+#RUN	git clone https://git.rockylinux.org/staging/rpms/qemu-kvm.git &&\
+#	git -C qemu-kvm checkout r8s-stream-rhel &&\
+#	if [[ ! -d /root/rpmbuild ]]; then mkdir /root/rpmbuild; fi &&\
+#	mv qemu-kvm/{SOURCES,SPECS} /root/rpmbuild/
+RUN sed -i -e 's/--disable-virtfs/--enable-virtfs/' \
+	-e 's/--disable-virtiofsd/--enable-virtiofsd/g' \
+	/root/rpmbuild/SPECS/qemu-kvm.spec
+RUN sed -i -e '/^%files -n qemu-kvm-common/,/^$/s/^$/%{_bindir}\/virtfs-proxy-helper\n%{_mandir}\/man1\/virtfs-proxy-helper.1.gz\n/' \
+    -e '/^%if %{rhev}$/,/^%else$/s/pkgsuffix -ev/pkgsuffix -virtfs/' \
+    -e '/%define rhel_rhev_conflicts()/ a Provides: %1-ev = %{epoch}:%{version}-%{release} \\\nObsoletes: %1-ev < %{obsoletes_version} \\' \
+    -e 's/rm -rf ${RPM_BUILD_ROOT}%{_mandir}\/man1\/virtfs-proxy-helper\*//g' \
+    -e 's/rm -rf ${RPM_BUILD_ROOT}%{_libexecdir}\/virtfs-proxy-helper/mv ${RPM_BUILD_ROOT}%{_libexecdir}\/virtfs-proxy-helper ${RPM_BUILD_ROOT}%{_bindir}\/virtfs-proxy-helper/g' \
+    /root/rpmbuild/SPECS/qemu-kvm.spec
+COPY 	virtio-9p-pci.patch /root/rpmbuild/SOURCES/
+COPY 	qemu-kvm.spec.patch /tmp
+RUN 	patch /root/rpmbuild/SPECS/qemu-kvm.spec /tmp/qemu-kvm.spec.patch
+#RUN	mkdir /tmp/qemu-6.2.0 &&\
+#	cp /root/rpmbuild/SOURCES/* /tmp/qemu-6.2.0/ &&\
+#	cd /tmp &&\
+#	tar cJf /tmp/qemu-6.2.0.tar.xz qemu-6.2.0 &&\
+#	mv /tmp/qemu-6.2.0.tar.xz /root/rpmbuild/SOURCES/
+RUN 	rpmbuild -ba --clean /root/rpmbuild/SPECS/qemu-kvm.spec
+
+# Build GNU telnetd server since classical builds do not allow running bash as login util
+RUN	wget "https://ftp.gnu.org/gnu/inetutils/inetutils-2.4.tar.xz" &&\
+	tar xf inetutils-2.4.tar.xz &&\
+	cd inetutils-2.4 &&\
+	./configure --disable-servers --disable-clients --enable-telnetd --enable-telnet &&\
+	make -j$(nproc)
+
+# =====================
+# KVM base container
+# =====================
+FROM 	${REGISTRY}/inginious/env-base:${VERSION}
```

Comment:
Two `FROM` directives in the same Dockerfile ?

Reply:
This is a multi-stage build (https://docs.docker.com/build/building/multi-stage/).
The first stage generates a patched QEMU RPM with P9FS enabled and the second stage is the actual container.

Reply:
The first stage also builds the GNU implementation of telnet which is not packaged for RL8.